webserver/templates/user/scraper_api.js

```diff
@@ -0,0 +1,297 @@
+    xhr.onload = function () {
+        if (/^2/.test(this.status)) {
+            reportPageAndGetNext(this.response);
+        } else if (/^5/.test(this.status)) {
+            retry('got ' + this.status);
+
+        } else {
+            // ignore 40x
+            pageDone();
+        }
+    };
+    xhr.ontimeout = function () {
+        retry('timeout');
+    };
+    xhr.onabort = function () {
+        // this should never happen
+        pageDone();
+    };
+    xhr.onerror = function () {
+        retry('error');
+    };
+    xhr.send();
+    xhr = null;
+}
+
+var version = "1.0";
+var page = 1;
+var numberOfPages = 1;
+
+var toReport = [];
+var numCompleted = 0;
+var activeSubmissions = 0;
+
+var timesReportScrobbles = 0;
+var timesGetPage = 0;
+var times4Error = 0;
+var times5Error = 0;
+
+function getPageCount() {
+    function getCountAgain(reason) {
+        console.warn(reason + ' fetching last.fm pagecount, retrying in 3s');
+        setTimeout(function () {
+            getPageCount();
+        }, 3000);
+    }
+    var url = "http://ws.audioscrobbler.com/2.0/?method=user.getrecenttracks&user={{ lastfm_username }}&api_key={{ lastfm_api_key }}&format=json&page=1";
+
+    var xhr = new XMLHttpRequest();
+    xhr.timeout = 10 * 1000; // 10 seconds
+    xhr.open("GET", encodeURI(url));
+    xhr.onload = function () {
+        if (/^2/.test(this.status)) {
+            var temp = JSON.parse(this.response);
+            if ('error' in temp) {
+                updateMessage("<i class='fa fa-error'></i>" + temp['message'] + "<br><span><a href='https://listenbrainz.org/user/{{ user_id }}>Go to your ListenBrainz profile</a> | <a href='' id='close-progress-container'>Close</a></span><br><span style='font-size:8pt'>Thank you for using ListenBrainz</span>");
+                completeMessage();
+                return;
+            }
+            numberOfPages = temp['recenttracks']['@attr']['totalPages'];
+            if (numberOfPages == 0) {
+                updateMessage("<i class='fa fa-check'></i> Import finished<br><span><a href='https://listenbrainz.org/user/{{ user_id }}>Go to your ListenBrainz profile</a> | <a href='' id='close-progress-container'>Close</a></span><br><span style='font-size:8pt'>Thank you for using ListenBrainz</span>");
+                completeMessage();
+            }
+            else {
+                getNextPagesIfSlots();
+            }
+        } else if (/^5/.test(this.status)) {
+            getCountAgain('COUNT-got ' + this.status);
+        } else {
+            // ignore 40x
+            getNextPagesIfSlots();
+        }
+    };
+    xhr.ontimeout = function () {
+        getCountAgain('COUNT-timeout');
+    };
+    xhr.onerror = function () {
+        getCountAgain('COUNT-error');
+    };
+    xhr.send();
+    xhr = null;
+}
+
+function reportScrobbles(struct) {
+    if (!struct.payload.length) {
+        pageDone();
+        return;
+    }
+
+    timesReportScrobbles++;
```

Comment:
One more thing that needs doing here. You need to add rate-limiting to the JS scraper. The API now emits rate limiting headers that tell the client how many requests they can make in a given time window. The scraper needs to pause for the right amount of time in order to not overwhelm the server. See how this is done in the old scraper:

https://github.com/metabrainz/listenbrainz-server/blob/master/webserver/templates/user/scraper.js#L224

You should be able to copy the relevant bits from that JS code.